src/openrct2/actions/GameAction.h

```diff
@@ -140,6 +140,38 @@ class GameActionResult
     std::string GetErrorMessage() const;
 };
 
+/**
+ *
+ */
+class GameActionParameterVisitor
+{
+public:
+    virtual ~GameActionParameterVisitor() = default;
+
+    virtual void Visit(const std::string_view& name, int32_t& param)
+    {
+    }
+
+    virtual void Visit(const std::string_view& name, std::string& param)
+    {
+    }
+
+    void Visit(CoordsXYZD& param)
+    {
+        Visit("x", param.x);
+        Visit("y", param.y);
+        Visit("z", param.z);
+        Visit("direction", param.direction);
+    }
+
+    template<typename T> void Visit(const std::string_view& name, T& param)
+    {
+        auto value = static_cast<int32_t>(param);
```

Comment:
Maybe we should assert that it is number-like, to make compilation errors clearer, should we ever try to visit it with a `T` that is not
```suggestion
        static_assert(std::is_arithmetic<T>::value, "Not an arithmetic type");
        auto value = static_cast<int32_t>(param);
```